### PR TITLE
Identifier handling in ICS02, 03, 04 (+ some cleanup)

### PR DESCRIPTION
**Pull request by adizere**
Closes: #622 

## Description

______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [x] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/ics04_channel/context.rs [hunk 1]
```diff
@@ -20,51 +21,68 @@ pub trait ChannelReader {
...
+                    &(result.port_id.clone(), result.channel_id.clone()),
+                    &result.channel_end,
+                )?;
+
+                // Associate also the channel end to its connection.
+                self.store_connection_channels(
+                    &result.channel_end.connection_hops()[0].clone(),
+                    &(result.port_id.clone(), result.channel_id.clone()),
+                )?;
+
```
Comment (ancazamfir) [inline modules/src/ics04_channel/context.rs:59]: I liked it better before :) I don't think we should read too deeply into the channel state for everything.  We should not be trying to validate what `process` did here (*). Same for connection and I provided similar comments there. 
What happens for `State::TryOpen`? don't we have a similar case as for Init?
I think we should just store the channel and store it under connection regardless of state. Increase the channel counter if previous_channel_id is None. Store the sequences only for `Init` state. 

(*) if we do, then define different result types for each message.

File: modules/src/ics04_channel/handler/chan_close_confirm.rs [hunk 2]
```diff
@@ -82,7 +83,8 @@ pub(crate) fn process(
 
     let result = ChannelResult {
         port_id: msg.port_id().clone(),
-        channel_id: Some(msg.channel_id().clone()),
+        channel_id: msg.channel_id().clone(),
+        previous_channel_id: None,
```
Comment (ancazamfir) [inline modules/src/ics04_channel/handler/chan_close_confirm.rs:87]: Why `None`? Same comment for most of the other handlers.

File: modules/src/ics04_channel/handler/chan_open_confirm.rs [hunk 3]
```diff
@@ -131,73 +127,43 @@ mod tests {
...
         let conn_end = ConnectionEnd::new(
             ConnectionState::Open,
-            msg_conn_try.client_id().clone(),
-            ConnectionCounterparty::new(
-                msg_conn_try.counterparty().client_id().clone(),
-                Some(ConnectionId::from_str("defaultConnection-0").unwrap()),
-                msg_conn_try.counterparty().prefix().clone(),
-            ),
+            client_id.clone(),
+            ConnectionCounterparty::try_from(get_dummy_raw_counterparty()).unwrap(),
```
Comment (ancazamfir) [inline modules/src/ics04_channel/handler/chan_open_confirm.rs:139]: We should have the same signature for `get_dummy_raw_counterparty()` as we have for channel.

**Review by ancazamfir — CHANGES_REQUESTED** [review]
Great work @Adi! A lot of good changes. Not your fault, I guess it's just the nature of the beast (many messages with many fields and many corner cases to test) but the unit tests are not easy to follow in general and hard to maintain. Just did a cursory review there. Hope with MBT things will be better

File: modules/src/ics03_connection/msgs/conn_open_try.rs [hunk 4]
```diff
@@ -76,6 +76,11 @@ impl MsgConnectionOpenTry {
         }
     }
 
+    /// Setter for `client_id`.
+    pub fn with_client_id(self, client_id: ClientId) -> MsgConnectionOpenTry {
+        MsgConnectionOpenTry { client_id, ..self }
+    }
+
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/ics03_connection/msgs/conn_open_try.rs]: What about this approach 32732c29 ?

File: modules/src/ics03_connection/context.rs [hunk 5]
```diff
@@ -62,36 +67,35 @@ pub trait ConnectionKeeper {
     fn store_connection_result(&mut self, result: ConnectionResult) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/ics03_connection/context.rs:67]: Change with a different (simpler) strategy.

File: modules/src/ics03_connection/msgs/conn_open_try.rs [hunk 6]
```diff
@@ -93,13 +98,13 @@ impl Msg for MsgConnectionOpenTry {
...
-        vec![self.signer]
-    }
-
     fn type_url(&self) -> String {
         TYPE_URL.to_string()
     }
+
+    fn get_signers(&self) -> Vec<AccountId> {
+        vec![self.signer]
+    }
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/ics03_connection/msgs/conn_open_try.rs:91]: Yep, that could be a next priority. (ref: https://github.com/informalsystems/ibc-rs/issues/219#issuecomment-724715856)

File: modules/src/ics04_channel/context.rs (hunk 1 above)
Comment (adizere) [reply modules/src/ics04_channel/context.rs]: Refactor with a simpler approach.

File: modules/src/ics03_connection/msgs/conn_open_try.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cezarad) [reply modules/src/ics03_connection/msgs/conn_open_try.rs]: We have tests for the other message types/handler but we modify "manually" the default message fields, to match the tests needs.

File: modules/src/ics04_channel/channel.rs [hunk 7]
```diff
@@ -113,6 +113,7 @@ impl ChannelEnd {
     pub fn set_state(&mut self, s: State) {
         self.state = s;
     }
+
     pub fn set_version(&mut self, v: String) {
         self.version = v;
     }
```
Comment (cezarad) [inline modules/src/ics04_channel/channel.rs:119]: could you please add 

```
pub fn set_counterparty_channel_id(&mut self, c: ChannelId) {
        self.remote.channel_id = Some(c);
    }
```


Reading I realized it's required in chan_open_ack

File: modules/src/ics04_channel/handler/chan_open_ack.rs [hunk 8]
```diff
@@ -84,7 +84,8 @@ pub(crate) fn process(
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/ics04_channel/handler/chan_open_ack.rs:85]: Great catch! fcf6c371

File: modules/src/ics04_channel/handler/chan_open_ack.rs [hunk 9]
```diff
@@ -74,7 +74,7 @@ pub(crate) fn process(
         &expected_channel_end,
         &msg.proofs(),
     )
-    .map_err(|e| Kind::FailedChanneOpenAckVerification.context(e))?;
+    .map_err(|e| Kind::ChanOpenAckProofVerification.context(e))?;
 
     output.log("success: channel open try ");
```
[the remark this answers is not in the document]
Comment (adizere) [reply modules/src/ics04_channel/handler/chan_open_ack.rs]: fcf6c371

File: modules/src/ics04_channel/channel.rs (hunk 7 above)
Comment (adizere) [reply modules/src/ics04_channel/channel.rs:119]: Thanks, should be fixed in fcf6c371.

**Review by cezarad — APPROVED** [review]

File: modules/src/ics04_channel/handler/chan_open_confirm.rs (hunk 3 above)
Comment (adizere) [reply modules/src/ics04_channel/handler/chan_open_confirm.rs:139]: Can you expand on this suggestion? `get_dummy_raw_channel_end` has similar signature with `get_dummy_raw_counterparty`.

Comment (ancazamfir) [reply modules/src/ics04_channel/handler/chan_open_confirm.rs:139]: looks good now, i was referring to channel vs connection.

**Review by ancazamfir — APPROVED** [review]
Looks great!!! Thanks @Adi!

**Comment by adi** [comment]
Hi Anca, all,

Please remove me (@adi) from your mailing list as I'm certainly not the guy
you're looking for.

Mult spor în ceea faceți!
Alt Adi
